File: php/src/Advertisement/Application/Command/UpdateAdvertisement/UpdateAdvertisementUseCase.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisement\Application\Command\UpdateAdvertisement;

use Demo\App\Advertisement\Domain\AdvertisementRepository;
use Demo\App\Advertisement\Domain\Model\ValueObject\Password;
use Exception;

final class UpdateAdvertisementUseCase
{
    public function __construct(private AdvertisementRepository $advertisementRepository)
    {
    }

    /**
     * @throws Exception
     */
    public function execute(UpdateAdvertisementCommand $command): void
    {
        $advertisement = $this->advertisementRepository->findById($command->id);

        if (!$advertisement->password()->isValidatedWith($command->password)) {
            return;
        }

        $advertisement->update($command->description, Password::fromPlainPassword($command->password));

        $this->advertisementRepository->save($advertisement);
    }
}

File: kotlin/src/main/kotlin/advertisement/application/updateAdvertisement/UpdateAdvertisementUseCase.kt
package advertisement.application.updateAdvertisement

import advertisement.domain.AdvertisementRepository
import advertisement.domain.model.value_object.Password
import advertisement.infrastructure.PasswordHasher

class UpdateAdvertisementUseCase(private val advertisementRepository: AdvertisementRepository, private val passwordHasher: PasswordHasher) {
    fun execute(updateAdvertisementCommand: UpdateAdvertisementCommand) {
        val advertisement = advertisementRepository.findById(updateAdvertisementCommand.id)

        if (!advertisement.password?.validateWith(updateAdvertisementCommand.password, passwordHasher)!!)
            return

        advertisement.update(
            updateAdvertisementCommand.description,
            Password.fromPlainPassword(updateAdvertisementCommand.password, passwordHasher)
        )

        advertisementRepository.save(advertisement)
    }
}
